Ruby: Rework call graph implementation

hvitved opened the pull request:
This PR rewrites the call graph computation in the following ways:
1. Remove temporary fix for `require` and `require_relative` methods.
2. No longer resolve instance methods based on flow through `self` (closed-world assumption); instead rely on the type hierarchy, and allow dispatch to any overrides (open-world assumption), as long as the type is not known exactly (such as when constructed with `new`).
3. Improve precision for singleton methods defined on objects, such as `def x.foo; end`, by taking overriding redefinitions into account:
```rb
def x.foo; end
def x.foo; end
x.foo # now only resolves to the latter definition
```
5. Add missing call targets for singleton methods defined on parameters (including `self`):
```rb
def add_singleton x
  def x.singleton; end
end
add_singleton y
y.singleton # now has a call target
```
7. Add missing call targets for instance methods on objects referenced by variables introduced via pattern matching:
```rb
case object
  in C => c then c.foo # resolves in the same was as a call to `foo` inside `C`
end
```
9. Take type information  into account for objects that are being type checked:
```rb
case object
  when C then object.foo # resolves in the same was as a call to `foo` inside `C`
end
```

Commit-by-commit review is encouraged.

I spot checked some of the missing results on `opal/opal`, and they look like false positives to me. We were previously generating a lot of flow through [this](https://github.com/ruby/mspec/blob/098b3200ac8925a0285c6ccc29089e1e19908a07/lib/mspec/utils/format.rb#L9) call (included as a sub-module), which now has only 1 call target instead of 42, due to the change 2 above.

asgerf commented on ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:
```diff
@@ -293,71 +302,178 @@ private module Cached {
+ * The Boolean `exact` indicates whether `n` refers to exactly `m` (and not
+ * possibly a sub class).
+ */
+private predicate resolveModuleNode(DataFlow::Node n, boolean exact, Module m) {
```
As question as above

asgerf commented on ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:
```diff
@@ -293,71 +302,178 @@ private module Cached {
-    result.asExpr().getExpr() instanceof NilLiteral and tp = TResolved("NilClass")
+    result.asExpr().getExpr() instanceof NilLiteral and
+    tp = TResolved("NilClass") and
+    exact = false
```
Was this supposed to say `exact = true` here?

hvitved replied:
I guess so, yes.

asgerf reviewed (commented):
Could you do a DCA run with meta-queries turned on, so we can see the change in call graph edges?

hvitved commented on ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:
```diff
@@ -293,71 +302,178 @@ private module Cached {
+ * possibly a sub class).
+ */
+private predicate resolveModule(Expr e, boolean exact, Module m) {
+  m = resolveConstantReadAccess(e) and
+  exact = true
+  or
+  selfHasType(e.(SelfVariableReadAccess).getVariable(), exact, m)
```
Thanks for pointing out; this is indeed a bit confusing, and I will try to make the distinction more clear.

hvitved replied:
I have pushed an update. In making this rewrite, I also realized that two cases weren't handled correctly, as witnessed by updated test output on the second commit.

asgerf reviewed (commented):
Overall I think it's worth devoting a bit more time to the investigation of the call graph changes from the affected projects. Obviously the changes are too large to review in full. but normally I'd pick a few samples from each project and triage those. It's not really clear to me how much triaging has been done here.

hvitved commented on ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:
```diff
@@ -293,73 +425,156 @@ private module Cached {
+    // methods cannot be called; only singleton methods)
+    result =
+      any(SsaSelfDefinitionNode self |
+        selfInMethod(self.getVariable(), tp) and
```
We have [this logic](https://github.com/github/codeql/blob/ac4d4ff613ad5d9be6044e34d425d09bf9bfe050/ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll#L310-L315) in place for restricting calls to top-level methods, but I guess we also need a check the other way around, which should catch the example above.

hvitved replied:
Nevermind my comment above; I see now what you mean.

hvitved replied:
Another good catch, btw.

hvitved commented on ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:
```diff
@@ -293,73 +425,156 @@ private module Cached {
+      selfInModule(result.(SsaSelfDefinitionNode).getVariable(), m)
+      or
+      // needed for e.g. `self.puts`
+      selfInMethod(result.(SsaSelfDefinitionNode).getVariable(), m)
```
Good catch.

hvitved commented:
@asgerf: Thanks for the review. I have now spot checked two new call graph edges for each project, and here is what I found:

| Project | Call | Target | Conclusion |
|--------|--------|--------|--------|
| `activeadmin__activeadmin` | [call](https://github.com/activeadmin/activeadmin/blob/4f1ef18bd720ca272025735e769564104ccb4842/lib/active_admin/csv_builder.rb#L25) | [target](https://github.com/activeadmin/activeadmin/blob/4f1ef18bd720ca272025735e769564104ccb4842/lib/active_admin/csv_builder.rb#L41) | Looks actually like a TP, but it is resolved for incorrect reasons. I have pushed a fix. |
| `activeadmin__activeadmin` | [call](https://github.com/activeadmin/activeadmin/blob/4f1ef18bd720ca272025735e769564104ccb4842/lib/active_admin/views/pages/base.rb#L19) | [target](https://github.com/activeadmin/activeadmin/blob/4f1ef18bd720ca272025735e769564104ccb4842/lib/active_admin/views/pages/form.rb#L8) | Open-world |
| `aws__aws-sdk-ruby` | [call](https://github.com/aws/aws-sdk-ruby/blob/8a944fc07d9dcd454ee7727145e7514fa8ade27c/build_tools/customizations.rb#L44) | [target](https://github.com/aws/aws-sdk-ruby/blob/8a944fc07d9dcd454ee7727145e7514fa8ade27c/build_tools/customizations.rb#L38) | Open-world |
| `aws__aws-sdk-ruby` | [call](https://github.com/aws/aws-sdk-ruby/blob/8a944fc07d9dcd454ee7727145e7514fa8ade27c/build_tools/wrap_list.rb#L20) | [target](https://github.com/aws/aws-sdk-ruby/blob/8a944fc07d9dcd454ee7727145e7514fa8ade27c/build_tools/wrap_list.rb#L33) | Singleton class defined on `self` |
| `backup__backup` | [call](https://github.com/backup/backup/blob/69b25d7be4017a9dc1bddf85eaa8c93f78f327ce/lib/backup/utilities.rb#L211) | [target](https://github.com/backup/backup/blob/69b25d7be4017a9dc1bddf85eaa8c93f78f327ce/lib/backup/utilities.rb#L119) | Singleton class defined on `self` |
| `backup__backup` | [call](https://github.com/backup/backup/blob/69b25d7be4017a9dc1bddf85eaa8c93f78f327ce/spec/support/sandbox_file_utils.rb#L240) | [target](https://github.com/backup/backup/blob/69b25d7be4017a9dc1bddf85eaa8c93f78f327ce/spec/support/sandbox_file_utils.rb#L251) | Singleton class defined on `self` |
| `diaspora__diaspora` | [call](https://github.com/diaspora/diaspora/blob/1c064e016f4d1d1657778fcc739df8cd25919d76/app/controllers/application_controller.rb#L59) | [target](https://github.com/diaspora/diaspora/blob/1c064e016f4d1d1657778fcc739df8cd25919d76/app/controllers/api/openid_connect/user_info_controller.rb#L23) | Open-world |
| `diaspora__diaspora` | [call](https://github.com/diaspora/diaspora/blob/1c064e016f4d1d1657778fcc739df8cd25919d76/app/mailers/notification_mailers/base.rb#L21) | [target](https://github.com/diaspora/diaspora/blob/1c064e016f4d1d1657778fcc739df8cd25919d76/app/mailers/notification_mailers/also_commented.rb#L8) | Open-world |
| `discourse__discourse ` | [call](https://github.com/discourse/discourse/blob/c1887d95ffd749df86d55acd6630e8ece85614d1/app/controllers/bootstrap_controller.rb#L57) | [target](https://github.com/discourse/discourse/blob/c1887d95ffd749df86d55acd6630e8ece85614d1/app/controllers/extra_locales_controller.rb#L47) | Singleton method on module |
| `discourse__discourse ` | [call](https://github.com/discourse/discourse/blob/c1887d95ffd749df86d55acd6630e8ece85614d1/app/models/concerns/reports/posts.rb#L17) | [target](https://github.com/discourse/discourse/blob/c1887d95ffd749df86d55acd6630e8ece85614d1/app/models/category.rb#L173) | Singleton method on module |
| `errbit__errbit` | [call](https://github.com/doorkeeper-gem/doorkeeper/blob/bfc132f58455052db7b0c7f936fb72ffd033130d/lib/doorkeeper/oauth/base_request.rb#L15) | [target](https://github.com/doorkeeper-gem/doorkeeper/blob/bfc132f58455052db7b0c7f936fb72ffd033130d/lib/doorkeeper/oauth/client_credentials_request.rb#L19) | Open-world |
| `errbit__errbit` | [call](https://github.com/doorkeeper-gem/doorkeeper/blob/bfc132f58455052db7b0c7f936fb72ffd033130d/lib/doorkeeper/oauth/token.rb#L49) | [target](https://github.com/doorkeeper-gem/doorkeeper/blob/bfc132f58455052db7b0c7f936fb72ffd033130d/lib/doorkeeper/oauth/token.rb#L52) | Singleton class defined on `self` |
| `fastlane__fastlane` | [call](https://github.com/fastlane/fastlane/blob/ff41f3556f1baf8e700477e50f65be728780aa94/fastlane_core/lib/fastlane_core/ios_app_identifier_guesser.rb#L66) | [target](https://github.com/fastlane/fastlane/blob/ff41f3556f1baf8e700477e50f65be728780aa94/fastlane_core/lib/fastlane_core/ios_app_identifier_guesser.rb#L75) | Singleton class defined on `self` |
| `fastlane__fastlane` | [call](https://github.com/fastlane/fastlane/blob/ff41f3556f1baf8e700477e50f65be728780aa94/fastlane/lib/fastlane/actions/increment_version_number.rb#L42) | [target](https://github.com/fastlane/fastlane/blob/ff41f3556f1baf8e700477e50f65be728780aa94/fastlane/lib/fastlane/actions/increment_version_number.rb#L91) | Singleton methods in same class |
| `gitlabhq__gitlabhq` | [call](https://github.com/gitlabhq/gitlabhq/blob/5947e68ac310a25614438df36149b27b0654eb5c/app/controllers/import/base_controller.rb#L68) | [target](https://github.com/gitlabhq/gitlabhq/blob/5947e68ac310a25614438df36149b27b0654eb5c/app/controllers/import/manifest_controller.rb#L79) | Open-world |
| `gitlabhq__gitlabhq` | [call](https://github.com/gitlabhq/gitlabhq/blob/5947e68ac310a25614438df36149b27b0654eb5c/lib/gitaly/server.rb#L18) | [target](https://github.com/gitlabhq/gitlabhq/blob/5947e68ac310a25614438df36149b27b0654eb5c/lib/gitaly/server.rb#L9) | Singleton class defined on `self` |
| `googleapis__google-cloud-ruby` | [call](https://github.com/googleapis/google-cloud-ruby/blob/024d05cddb39fc98af14791816bf5c181d034a9c/google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb#L189) | [target](https://github.com/googleapis/google-cloud-ruby/blob/024d05cddb39fc98af14791816bf5c181d034a9c/google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb#L3047) | Open-world |
| `googleapis__google-cloud-ruby` | [call](https://github.com/googleapis/google-cloud-ruby/blob/024d05cddb39fc98af14791816bf5c181d034a9c/google-cloud-bigtable/lib/google/cloud/bigtable/service.rb#L284) | [target](https://github.com/googleapis/google-cloud-ruby/blob/024d05cddb39fc98af14791816bf5c181d034a9c/google-cloud-bigtable/test/helper.rb#L36) | Open-world (mock override)|
| `hahwul__mad-metasploit` | [call](https://github.com/hahwul/mad-metasploit/blob/34bde59d38e735050a4bfb263bf5afa476ec4ed1/mad-metasploit-plugins/meta_ssh/lib/net/sftp/protocol/01/attributes.rb#L63) | [target](https://github.com/hahwul/mad-metasploit/blob/34bde59d38e735050a4bfb263bf5afa476ec4ed1/mad-metasploit-plugins/meta_ssh/lib/net/sftp/protocol/04/attributes.rb#L48) | Open-world (overridden singleton method). I have added a test cases for this, and also made sure that even more cases are covered. |
| `hahwul__mad-metasploit` | [call](https://github.com/hahwul/mad-metasploit/blob/34bde59d38e735050a4bfb263bf5afa476ec4ed1/mad-metasploit-plugins/meta_ssh/lib/net/sftp/protocol/base.rb#L30) | [target](https://github.com/hahwul/mad-metasploit/blob/34bde59d38e735050a4bfb263bf5afa476ec4ed1/mad-metasploit-plugins/meta_ssh/lib/net/sftp/protocol/01/base.rb#L30) | Open-world (overridden singleton method) |
| `heartcombo__devise` | [call](https://github.com/heartcombo/devise/blob/6d32d2447cc0f3739d9732246b5a5bde98d9e032/app/controllers/devise_controller.rb#L183) | [target](https://github.com/heartcombo/devise/blob/6d32d2447cc0f3739d9732246b5a5bde98d9e032/test/controllers/inherited_controller_i18n_messages_test.rb#L14) | Open-world (override in test code) |
| `heartcombo__devise` | [call](https://github.com/heartcombo/devise/blob/6d32d2447cc0f3739d9732246b5a5bde98d9e032/lib/devise/failure_app.rb#L107) | [target](https://github.com/heartcombo/devise/blob/6d32d2447cc0f3739d9732246b5a5bde98d9e032/test/failure_app_test.rb#L26) | Open-world (override in test code) |
| `jeremyevans__sequel` | [call](https://github.com/jeremyevans/sequel/blob/7a70ac6d719c21bcb404adf6159cbd2769d9246f/lib/sequel/adapters/ado.rb#L272) | [target](https://github.com/jeremyevans/sequel/blob/7a70ac6d719c21bcb404adf6159cbd2769d9246f/lib/sequel/dataset/actions.rb#L150) | Open-world (`yield` call resolves, because the base call now resolves) |
| `jeremyevans__sequel` | [call](https://github.com/jeremyevans/sequel/blob/7a70ac6d719c21bcb404adf6159cbd2769d9246f/lib/sequel/database/schema_methods.rb#L146) | [target](https://github.com/jeremyevans/sequel/blob/7a70ac6d719c21bcb404adf6159cbd2769d9246f/lib/sequel/adapters/shared/db2.rb#L104) | Open-world. Happens because the [`DataBaseMethods`](https://github.com/jeremyevans/sequel/blob/7a70ac6d719c21bcb404adf6159cbd2769d9246f/lib/sequel/adapters/shared/db2.rb#L10) is [included into](https://github.com/jeremyevans/sequel/blob/7a70ac6d719c21bcb404adf6159cbd2769d9246f/lib/sequel/adapters/ibmdb.rb#L177) the `Sequel::IBMDB::Database` module, which extends `Sequel::Database`  |

The general pattern is that we now find more due to the open-world assumption, which is not surprising, and it looks like tests code/mocks are a natural source of potential false positives, but I know we have the same issue in C#.

Triaging the results did reveal a couple of bugs around singleton methods, which I have now fixed.

asgerf reviewed (approved):
Thanks! This looks great 👍
